src/spyglass/utils/dj_mixin.py

```diff
@@ -535,3 +536,91 @@ def super_delete(self, *args, **kwargs):
         logger.warning("!! Using super_delete. Bypassing cautious_delete !!")
         self._log_use(start=time(), super_delete=True)
         super().delete(*args, **kwargs)
+
+    from spyglass.utils.dj_merge_tables import Merge
+
+    def restrict_from_upstream(self, key, **kwargs):
+        """Recursive function to restrict a table based on secondary keys of upstream tables"""
+        return restrict_from_upstream(self, key, **kwargs)
+
+
+def restrict_from_upstream(table, key, max_recursion=3):
+    """Recursive function to restrict a table based on secondary keys of upstream tables"""
+    print(f"table: {table.full_table_name}, key: {key}")
+    # Tables not to recurse through because too big or central
+    blacklist = [
+        "`common_nwbfile`.`analysis_nwbfile`",
+    ]
+
+    # Case: MERGE table
+    if (table := table & key) and max_recursion:
+        if isinstance(table, Merge):
+            parts = table.parts(as_objects=True)
+            restricted_parts = [
+                restrict_from_upstream(part, key, max_recursion - 1)
+                for part in parts
+            ]
+            # only keep entries from parts that got restricted
+            restricted_parts = [
+                r_part.proj("merge_id")
+                for r_part, part in zip(restricted_parts, parts)
+                if (
+                    not len(r_part) == len(part)
+                    or check_complete_restrict(r_part, key)
+                )
+            ]
+            # return the merge of the restricted parts
+            merge_keys = []
+            for r_part in restricted_parts:
+                merge_keys.extend(r_part.fetch("merge_id", as_dict=True))
+            return table & merge_keys
+
+        # Case: regular table
+        upstream_tables = table.parents(as_objects=True)
+        # prevent a loop where call Merge master table from part
+        upstream_tables = [
+            parent
+            for parent in upstream_tables
+            if not (
+                isinstance(parent, Merge)
+                and table.full_table_name in parent.parts()
+            )
+            and (parent.full_table_name not in blacklist)
+        ]
+        for parent in upstream_tables:
+            print(parent.full_table_name)
+            print(len(parent))
+            r_parent = restrict_from_upstream(parent, key, max_recursion - 1)
+            if len(r_parent) == len(parent):
+                continue  # skip joins with uninformative tables
+            table = safe_join(table, r_parent)
+            if check_complete_restrict(table, key) or not table:
+                print(len(table))
+                break
+    return table
+
+
+def check_complete_restrict(table, key):
```

Comment:
dj has a builtin for this called 'assert_join_compatibility' - if adopted, it would need to be wrapped in a try/except and call it using the key as a QueryExpression, but I'm inclined to let them handle this kind of thing rather than add more for us to maintain.